### PR TITLE
Add ability to define partitions in ignition file.

### PR DESCRIPTION
**Pull request by hhamid93**
On branch main
Your branch is up to date with 'origin/main'.

Changes to be committed:
        modified:   roles/patch_host_config/tasks/main.yml
	new file:   roles/patch_host_config/templates/patch-worker-partition.j2

### CODE REVIEW

File: roles/patch_host_config/tasks/main.yml [hunk 1]
```diff
@@ -93,3 +93,32 @@
           ]
         }
       register: http_reply
+
+- name: Load patch for worker partition
+  when: hostvars[host_name]['partitions'] is defined
+  block:
+    - name: Template patch partition
+      set_fact:
+        worker_partition: "{{ lookup('template', 'patch-worker-partition.j2') }}"
```
Comment (crwr45) [inline roles/patch_host_config/tasks/main.yml:102]: The template is going to need to account for the configuration in `hostvars[host_name]['partitions']`, as it's currently hard-coded to 10x1GiB partitions, which is not going to work for most cases.

File: roles/patch_host_config/templates/patch-worker-partition.j2 [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{
+	"ignition": {
+		"version": "3.2.0"
```
Comment (crwr45) [inline roles/patch_host_config/templates/patch-worker-partition.j2:3]: We are going to need to come up with a way to bypass this functionality for older version of OCP, as ignition spec 3.2 is not supported on 4.6 (and possibly 4.7). I suggest adding something to the inventory validation step, and documenting the `partitions` field in the sample inventory.

**Review by crwr45 — CHANGES_REQUESTED** [review]
This will be really useful to have baked in to Crucible rather than having to inject shell scripts to partition disks post-install.
There's a few changes to make to get it generic enough to be merged, though.

**Comment by hhamid93** [comment]
Makes sense! 
Will try to get the template corrected so that user is able to decide on the number of partition if required or not.

Making this work with an older version is highly unlikely since 3.1.0 & below ignition spec does not have the "resize" parameter for making this work. 

Need to further discuss this with crucible team before making another commit.

**Comment by mkochanowski** [comment]
Because the PR is still a work in progress, I converted it to a draft. 

Feel free to mark it as "ready for review" once it's finished. :)

File: docs/inventory.md [hunk 3]
```diff
@@ -405,6 +405,20 @@ The basic network configuration of the inventory for the fully bare metal deploy
...
+## Additional Partition Deployment
+
+Uncomment the lines below in the inventory.yml to set additional partitions if required on the nodes. Please provide the correct label and size(MiB).
+Add all additional paritions required under extra_partitions (examples for label:size provided below).
+Additional partitions can only be created for OCP 4.8+ deployments
+
+```yaml
+# ...
+disks:
+  - device: "{{ installation_path }}" #DO NOT CHANGE
```
Comment (nocturnalastro) [inline docs/inventory.md:417]: You might want to change that for instance if you have a second disk you wish to format

File: inventory.yml.sample [hunk 4]
```diff
@@ -291,11 +291,32 @@ all:
...
               # # Uncomment to set an alternate installation disk device for the node
               # installation_disk_path: /dev/sdb
-
+              
+              # # Uncomment to set additional partitions if required on nodes. Please provide the correct label and size(MiB).
+              # # The device can either be the drive in which RHCOS image needs to be installed or it can be any additional drive that requires new partitions.
+              # # Add all additional paritions required under extra_partitions (examples for label:size provided below).
+              # # Additional partitions can only be created for OCP 4.8+ deployments
+              # disks:
+              # - device: "{{ installation_path }}" #See above section
```
Comment (mkochanowski) [inline inventory.yml.sample:300]: Should `installation_path` be `installation_disk_path` instead?

This is a snippet from **docs/inventory.md**:
```yaml
disks:
  - device: "{{ installation_disk_path }}"
    extra_partitions:
      partition_1: 1024
      partition_2: 1024
 ```

File: roles/patch_host_config/defaults/main.yml [hunk 5]
```diff
@@ -2,3 +2,9 @@ ASSISTED_INSTALLER_BASE_URL: "{{ secure | ternary('https', 'http') }}://{{ hostv
 URL_ASSISTED_INSTALLER_CLUSTER: "{{ ASSISTED_INSTALLER_BASE_URL }}/clusters/{{ cluster_id }}"
 
 single_node_openshift_enabled: "{{ is_valid_single_node_openshift_config | default(false) }}"
+
+disks_rhcos_root: # Required rhcos_partion
+  label: root
+  number: 4
```
Comment (mkochanowski) [inline roles/patch_host_config/defaults/main.yml:8]: Does this mean that there will be 4 partitions?

File: inventory.yml.sample (hunk 4 above)
Comment (hhamid93) [reply inventory.yml.sample]: Good Catch!
This has been corrected in the inventory.yml.sample
Thanks Marek!

File: roles/patch_host_config/templates/patch-worker-partition.j2 [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    "version": "3.2.0" 
+  },
+  "storage": {
+    "disks": [
+      {% for disk in disks %}
+      {
+        "device": "{{ disk.device }}",
+        "partitions": [ 
+          {% set partitions = ((disk.extra_partitions | default({})).items() | list) %}
+          {% if disk.device == (installation_disk_path | default(host.json.installation_path)) %}
```
Comment (nocturnalastro) [inline roles/patch_host_config/templates/patch-worker-partition.j2:13]: @hhamid93 Did you check the json returned from the host endpoint to check this is the correct path?

File: roles/patch_host_config/tasks/main.yml [hunk 7]
```diff
@@ -95,3 +95,34 @@
...
+
+    - debug: # noqa unnamed-task
+        var: worker_partition
+        verbosity: 1
+  
+  
+    - name: Patch worker partition
+      uri:
+        url: "{{ URL_ASSISTED_INSTALLER_CLUSTER }}/hosts/{{host_id}}/ignition"
```
Comment (nocturnalastro) [inline roles/patch_host_config/tasks/main.yml:115]: You need to add spaces around host id thats why theres a warning on 113.

Comment (hhamid93) [reply roles/patch_host_config/tasks/main.yml]: Thanks ! 
This has been corrected

**Review by crwr45 — COMMENTED** [review]
Looking great!

Needs the commits squashing, and since the behaviour is only v4.8 or higher it would be great to have a version check for this. I suggest adding a task to inventory validation to fail if the version is too low and the partition data is provided.

File: inventory.yml.sample [hunk 8]
```diff
@@ -291,11 +291,32 @@ all:
...
+              # # Uncomment to set additional partitions if required on nodes. Please provide the correct label and size(MiB).
+              # # The device can either be the drive in which RHCOS image needs to be installed or it can be any additional drive that requires new partitions.
+              # # Add all additional paritions required under extra_partitions (examples for label:size provided below).
+              # # Additional partitions can only be created for OCP 4.8+ deployments
+              # disks:
+              # - device: "{{ installation_disk_path }}" #See above section
+              #   extra_partitions:
+              #     partition_1: 1024
+              #     partition_2: 102
+
```
Comment (arjuhe) [inline inventory.yml.sample:304]: This is an advanced feature and doesn't need to be in the sample inventory. The doc on how to use it should cover when, how, and where to use it.

File: inventory.yml.sample [hunk 9]
```diff
@@ -291,11 +291,32 @@ all:
...
+              # # Uncomment to set additional partitions if required on nodes. Please provide the correct label and size(MiB).
+              # # The device can either be the drive in which RHCOS image needs to be installed or it can be any additional drive that requires new partitions.
+              # # Add all additional paritions required under extra_partitions (examples for label:size provided below).
+              # # Additional partitions can only be created for OCP 4.8+ deployments
+              # disks:
+              # - device: "{{ installation_disk_path }}" #See above section
+              #   extra_partitions:
+              #     partition_1: 1024
+              #     partition_2: 1024
+
```
Comment (arjuhe) [inline inventory.yml.sample:322]: This is redundant here.

File: roles/patch_host_config/templates/patch-worker-partition.j2 [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+          {% endif %}
+          {% for label, size in partitions %}
+          {
+            "label": "{{ label | string }}",
+            "sizeMiB": {{ size | int }}
+          }{{ ((loop.index == (partitions | length)) | ternary('',',')) }}
+          {% endfor %}
+        ],
+        "wipeTable": false
+      }{{ ((loop.index == (disks | length)) | ternary('',',')) }}
```
Comment (arjuhe) [inline roles/patch_host_config/templates/patch-worker-partition.j2:24]: Should there be a new line between the curlies `}{{`?

**Review by arjuhe — CHANGES_REQUESTED** [review]
Nothing major. Almost there.

File: roles/patch_host_config/templates/patch-worker-partition.j2 (hunk 10 above)
Comment (nocturnalastro) [reply roles/patch_host_config/templates/patch-worker-partition.j2:24]: No we don't want a new line there as the comma that creates needs to be on end of the line loop is creating.

Comment (arjuhe) [reply roles/patch_host_config/templates/patch-worker-partition.j2:24]: 👍

File: inventory.yml.sample (hunk 8 above)
Comment (hhamid93) [reply inventory.yml.sample]: Change have been made. Removed from Sample inventory

File: inventory.yml.sample (hunk 9 above)
Comment (hhamid93) [reply inventory.yml.sample]: change have been made.

**Review by arjuhe — APPROVED** [review]

**Review by arjuhe — CHANGES_REQUESTED** [review]
> Looking great!
> 
> Needs the commits squashing, and since the behaviour is only v4.8 or higher it would be great to have a version check for this. I suggest adding a task to inventory validation to fail if the version is too low and the partition data is provided.

Please address the issue mentioned above.

**Comment by hhamid93** [comment]
> > Looking great!
> > Needs the commits squashing, and since the behaviour is only v4.8 or higher it would be great to have a version check for this. I suggest adding a task to inventory validation to fail if the version is too low and the partition data is provided.
> 
> Please address the issue mentioned above.

**Comment by hhamid93** [comment]
Validations have been added.

**Review by nocturnalastro — APPROVED** [review]
@arjuhe @crwr45 I think call comments have now been addressed

**Review by arjuhe — APPROVED** [review]
Congrats!!! +1

**Pull request merged**